templates/contest/list.html

```diff
@@ -124,7 +124,7 @@
 {% macro contest_join(contest, request) %}
     {% if not request.in_contest %}
         <td>
-            {% if request.profile in contest.authors.all() or request.profile in contest.curators.all() or request.profile in contest.testers.all() %}
+            {% if request.profile in contest.authors.all() or request.profile in contest.curators.all() or request.profile in contest.testers.all() or contest.has_completed_contest(request.user) %}
```

Comment:
This runs O(n) queries, where `n` is the number of active contests. I guess it's not that bad.

Reply:
I suppose the other option is to run a few SQL queries at the beginning so that it's just one SQL query that produces O(n) results. More complex obviously.

Reply:
The way participation end time is calculated is quite complex to do in a batch query (https://github.com/DMOJ/online-judge/blob/master/judge/models/contest.py#L471), so I think doing O(n) queries is the best option we have.

Reply:
Like Quantum said, it's probably not too bad. If the query is complex and the number of active contests is expected to be low, it's not worth changing this.

Reply:
> The way participation end time is calculated is quite complex to do in a batch query

`contests.py` already calculates `participation.ended` O(n) times. I pushed a commit that uses this calculation. There's no need to optimize `participation.ended`.